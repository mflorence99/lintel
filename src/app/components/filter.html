<div 
  class="filter">

  <input 
    (keyup)="filterRuleName($event)"
    placeholder="Filter rules">

</div>
