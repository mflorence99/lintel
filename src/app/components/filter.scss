:host {
  display: block;
}

.filter {
  background: var(--vscode-editorGroup-dropBackground);
  border: 1px solid var(--vscode-editorGroup-border);
  display: grid;
  grid-gap: 1rem;
  grid-template-columns: 1fr auto;
  padding: 0.5rem;

  fa-icon {
    color: var(--vscode-editorWidget-foreground);
    cursor: pointer;
    opacity: 0.6;
  }

  input {
    background: transparent;
    border: 0;
    color: var(--vscode-input-foreground);

    &:focus {
      outline: 0;
    }

  }

}
