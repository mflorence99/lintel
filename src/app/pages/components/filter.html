<form
  [formGroup]="filterForm"
  autocomplete="off"
  novalidate>

  <lintel-input
    formControlName="filter"
    placeholder="Filter rules">

    <fa-icon 
      (click)="toggleInheritedRules()"
      [icon]="['fas', filter.snapshot.showInheritedRules? 'eye' : 'eye-slash']"
      suffix
      title="Toggle inherited rules analysis">
    </fa-icon>

    <fa-icon 
      (click)="clearRuleNameFilter()"
      [icon]="['fas', 'times']"
      suffix
      title="Clear filter rules input">
    </fa-icon>

  </lintel-input>

</form>
