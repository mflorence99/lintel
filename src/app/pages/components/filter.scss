:host {
  display: block;
  margin-left: var(--rules-gutter);
}
