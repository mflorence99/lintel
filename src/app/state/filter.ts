import { DataAction } from '@ngxs-labs/data/decorators';
import { Debounce } from '@ngxs-labs/data/decorators';
import { Injectable } from '@angular/core';
import { NgxsDataRepository } from '@ngxs-labs/data/repositories';
import { Params } from '../services/params';
import { Payload } from '@ngxs-labs/data/decorators';
import { Persistence } from '@ngxs-labs/data/decorators';
import { State } from '@ngxs/store';
import { StateRepository } from '@ngxs-labs/data/decorators';

export type FilterCallback = () => void;

export interface FilterStateModel {
  ruleNameFilter?: string;
  showInheritedRules?: boolean;
}

@Injectable({ providedIn: 'root' })
@Persistence()
@StateRepository()
@State<FilterStateModel>({
  name: 'filter',
  defaults: { 
    showInheritedRules: true
  }
})

export class FilterState extends NgxsDataRepository<FilterStateModel> {

  // actions

  @DataAction({ insideZone: true })
  @Debounce(Params.debounceTimeout)
  filterRuleName(@Payload('filterRuleName') ruleNameFilter: string, done?: FilterCallback): void {
    this.ctx.patchState({ ruleNameFilter });
    done?.();
  }

  @DataAction({ insideZone: true })
  @Debounce(Params.debounceTimeout)
  toggleInheritedRules(done?: FilterCallback): void {
    const state = this.ctx.getState();
    this.ctx.patchState({ showInheritedRules: !state.showInheritedRules });
    done?.();
  }

}
