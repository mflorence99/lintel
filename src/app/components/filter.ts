import { ChangeDetectionStrategy } from '@angular/core';
import { Component } from '@angular/core';
import { ConfigsState } from '../state/configs';
import { ElementRef } from '@angular/core';
import { FilterCallback } from '../state/filter';
import { FilterState } from '../state/filter';
import { SelectionState } from '../state/selection';
import { ViewChild } from '@angular/core';

/**
 * Filter component
 */

@Component({
  changeDetection: ChangeDetectionStrategy.Default,
  selector: 'lintel-filter',
  templateUrl: 'filter.html',
  styleUrls: ['filter.scss']
})

export class FilterComponent {

  @ViewChild('input', { static: true }) input: ElementRef;

  /** ctor */
  constructor(public configs: ConfigsState,
              public filter: FilterState,
              public selection: SelectionState) { }

  /** Clear the rule name filter */
  clearRuleNameFilter(): void {
    this.filter.clearRuleNameFilter();
    this.input.nativeElement.value = null;
  }

  /** Filter rule names */
  filterRuleName(ruleNameFilter: string, done?: FilterCallback): void {
    this.filter.filterRuleName(ruleNameFilter, () => {
      // if the currently-selected category is now empty, reset selection
      if (!this.configs.categories.includes(this.selection.category))
        this.selection.select({ category: this.configs.categories[0] });
      done?.();
    });
  }

}
