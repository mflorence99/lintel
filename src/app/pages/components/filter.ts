import { ChangeDetectionStrategy } from '@angular/core';
import { Component } from '@angular/core';
import { ConfigsState } from '../../state/configs';
import { FilterState } from '../../state/filter';
import { FormBuilder } from '@angular/forms';
import { FormGroup } from '@angular/forms';
import { OnDestroy } from '@angular/core';
import { OnInit } from '@angular/core';
import { SelectionState } from '../../state/selection';
import { Subject } from 'rxjs';

import { takeUntil } from 'rxjs/operators';

/**
 * Filter component
 */

@Component({
  changeDetection: ChangeDetectionStrategy.Default,
  selector: 'lintel-filter',
  templateUrl: 'filter.html',
  styleUrls: ['filter.scss']
})

export class FilterComponent implements OnDestroy, OnInit {

  filterForm: FormGroup;

  private notifier = new Subject<void>();

  /** ctor */
  constructor(public configs: ConfigsState,
              public filter: FilterState,
              private formBuilder: FormBuilder,
              public selection: SelectionState) { 
    // create the filter form
    this.filterForm = this.formBuilder.group({
      filter: this.filter.snapshot.ruleNameFilter
    });
  }

  /** Clear the rule name filter */
  clearRuleNameFilter(): void {
    this.filter.filterRuleName(null);
    this.filterForm.setValue({ filter: null });
  }

  /** When we're done */
  ngOnDestroy(): void {
    this.notifier.next();
    this.notifier.complete();
  }

  /** When we're ready */
  ngOnInit(): void {
    this.filterForm.valueChanges
      .pipe(takeUntil(this.notifier))
      .subscribe(filterForm => this.filter.filterRuleName(filterForm.filter));
  }

  /** Show or hide inherited rules */
  toggleInheritedRules(): void {
    this.filter.toggleInheritedRules();
  }

}
