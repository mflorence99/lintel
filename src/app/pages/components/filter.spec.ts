import { FilterComponent } from './filter';

import { prepare } from './component.spec';

import { TestBed } from '@angular/core/testing';

import { async } from '@angular/core/testing';

describe('FilterComponent', () => {

  beforeEach(async(() => prepare()));

  test('Component is created', () => {
    const fixture = TestBed.createComponent(FilterComponent);
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });

  test('clearRuleNameFilter', () => {
    const fixture = TestBed.createComponent(FilterComponent);
    const component = fixture.componentInstance;
    component.filter.filterRuleName('xxx');
    expect(component.filter.snapshot.ruleNameFilter).toBe('xxx');
    component.clearRuleNameFilter();
    expect(component.filter.snapshot.ruleNameFilter).toBeNull();
  });

  test('ngOnInit', () => {
    const fixture = TestBed.createComponent(FilterComponent);
    const component = fixture.componentInstance;
    component.ngOnInit();
    component.filterForm.setValue({ filter: 'yyy' });
    expect(component.filter.snapshot.ruleNameFilter).toBe('yyy');
  });

  test('toggleInheritedRules', () => {
    const fixture = TestBed.createComponent(FilterComponent);
    const component = fixture.componentInstance;
    component.filter.showInheritedRules();
    expect(component.filter.snapshot.showInheritedRules).toBe(true);
    component.toggleInheritedRules();
    expect(component.filter.snapshot.showInheritedRules).toBe(false);
  });

});
