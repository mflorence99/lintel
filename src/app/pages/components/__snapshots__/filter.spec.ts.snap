// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterComponent snapshot 1`] = `
<lintel-filter
  __ngContext__={[Function LRootView]}
  configs={[Function ConfigsState]}
  destroy$={[Function DestroyService]}
  filter={[Function FilterState]}
  filterForm={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  params={[Function Params]}
  selection={[Function SelectionState]}
>
  <lintel-input
    class="ng-untouched ng-pristine ng-valid"
    formcontrolname="filter"
    placeholder="Filter rules"
  >
    <input
      autocomplete="off"
      max="undefined"
      min="undefined"
      placeholder="Filter rules"
      step="undefined"
      tabindex="0"
      type="text"
    />
    <fa-icon
      class="ng-fa-icon"
      suffix=""
      title="Toggle inherited rules analysis"
    >
      <svg
        aria-labelledby="svg-inline--fa-title-xxx"
        class="svg-inline--fa fa-eye fa-w-18"
        data-icon="eye"
        data-prefix="fas"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title
          id="svg-inline--fa-title-xxx"
        >
          Toggle inherited rules analysis
        </title>
        <path
          d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
          fill="currentColor"
        />
      </svg>
    </fa-icon>
    <fa-icon
      class="ng-fa-icon"
      suffix=""
      title="Clear filter rules input"
    >
      <svg
        aria-labelledby="svg-inline--fa-title-xxx"
        class="svg-inline--fa fa-times fa-w-11"
        data-icon="times"
        data-prefix="fas"
        role="img"
        viewBox="0 0 352 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title
          id="svg-inline--fa-title-xxx"
        >
          Clear filter rules input
        </title>
        <path
          d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          fill="currentColor"
        />
      </svg>
    </fa-icon>
  </lintel-input>
</lintel-filter>
`;
